codice/ddf
# DDF-4133 Solr auth updates

#### What does this PR do?
* Fixed formatting errors in documentation
* Improved logging and auditing
* Incorporated changes from Austin that got lost


#### Who is reviewing it? 
@austinsteffes 
@ricklarsen 
@paouelle 

#### Select relevant component teams: 
@codice/security 
@codice/solr 

#### Ask 2 committers to review/merge the PR and tag them here.
@clockard 
@vinamartin

#### How should this be tested?
This is how it was tested:
### TEST - DO NOT AUTOGENERATED PASSWORD
1. Set autogeneratePassword=false
2. Install
3. Log into Solr https://localhost:8994/solr
4. Username/password ddf/autogenerated 
5. Expected behavior
   * Solr login succeeeds
   * DF connects successfully to Solr cores
   * Alerts appear on Admin UI
   * Able to upload and search for a product
6. Actual behavior:
   *as expected
7. Test passed

### TEST - AUTOGENERATE PASSWORD
1. Set autogeneratedPassword=true
2. Restart
3. Expected behavior
   * Cannot log into Solr UI with ddf/autogenerated
   * DF connects successfully to Solr cores
   * Alerts appear on Admin UI
   * Able to upload and search for a product
   * Log messsages about password change exists
   * Security audit log messages about password change exists
4. Actual behavior
   * As expected
   * System log messages:
       * `2018-12-13T10:41:11,959 | INFO  | rClientAdapter 0 | SolrPasswordUpdateImpl           | .security.SolrPasswordUpdateImpl  166 | 374 - solr-password - 2.14.0.SNAPSHOT | "Updated encrypted Solr password in properties file /sandbox/ddf-2.14.0-SNAPSHOT/etc/custom.system.properties.`
       * `2018-12-13T10:41:11,960 | INFO  | rClientAdapter 0 | SolrPasswordUpdateImpl           | .security.SolrPasswordUpdateImpl  120 | 374 - solr-password - 2.14.0.SNAPSHOT | Updated encrypted Solr password in memory.`
       * `2018-12-13T10:41:11,942 | INFO  | rClientAdapter 0 | SolrPasswordUpdateImpl           | .security.SolrPasswordUpdateImpl  180 | 374 - solr-password - 2.14.0.SNAPSHOT | New password was set in Solr server.`
   * Security audit log messages:
       * `[INFO ] 2018-12-13T10:41:11,943 | rClientAdapter 0 | securityLogger  |  Subject: UNKNOWN New password was set in Solr server by the Solr password updater.`
       * `[INFO ] 2018-12-13T10:41:11,958 | rClientAdapter 0 | securityLogger  |  Subject: UNKNOWN "Updated encrypted Solr password in properties file /sandbox/ddf-2.14.0-SNAPSHOT/etc/custom.system.properties.`
6. Test passed
       
### TEST - MANUALLY RESTORE SOLR PASSWORD AND MANUALLY UPDATE SYSTEM PROPERTIES
1. Edit solr/server/solr/security.json and replace hashed password
2. Set autogeneratePassword to false in system properties file
3. Set solrPassword to "autogenerated" in system properties file
3. Start System
4. Expected behavior: DDF connects to Solr successfully
5. Actual behavior: as expected
7. Test passed

### NEGATIVE TEST - FAIL TO UPDATE SOLR PASSWORD
1. (From a clean distribution) Set start.solr to "false" in system properties file
5. Start System. DDF cannot communicate with Sorl because the Solr process was never started.
6. Expected behavior:
    * Solr password in system properties remains "autogenerated"
    * Message in log file that Solr password could not be updated
    * Message in audit log that Solr password could not be updated
7. Actual behavior: as expected
8. Test passed
    
### NEGATIVE TEST - FAIL TO UPDATE SYSTEM PROPERTIES FILE
1. (From a clean distribution) Set start.solr to "false" in system properties file
2. Put in a break point where DDF would save the new password in the system properties file
3. Start System with debugger attached.
4. When the break point is hit, remove all write permission from the custom.system.properites file. This will prevent DDF from updating the file.
3. Expected behavior:
   * Messages in log and audit file that Solr password was updated, but not saved.
   * DDF will not be able to communicate with Solr (because the Solr process was updated, but the new password was not saved so DDF does not know the new password)
4. Actual behavior:
    * `2018-12-13T15:19:45,081 | ERROR | rClientAdapter 0 | SolrPasswordUpdateImpl           | .security.SolrPasswordUpdateImpl  172 | 374 - solr-password - 2.14.0.SNAPSHOT | Exception while writing to /sandbox/ddf-2.14.0-SNAPSHOT/etc/custom.system.properties. Solr password was changed, but new password was not saved.`
    * `[INFO ] 2018-12-13T15:19:00,466 | rClientAdapter 0 | securityLogger  |  Subject: UNKNOWN New password was set in Solr server by the Solr password updater.`
5. Test passed

#### Any background context you want to provide?
No.

#### What are the relevant tickets?
[[[DDF-4133]](https://codice.atlassian.net/browse/DDF-4133)](https://codice.atlassian.net/browse/DDF-4133)

#### Checklist:
- [NA] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
